controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/UpdateReaderGroupTask.java

```diff
@@ -0,0 +1,94 @@
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataStore = streamMetaStore;
+        this.streamMetadataTasks = streamMetaTasks;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(final UpdateReaderGroupEvent request) {
+        String scope = request.getScope();
+        String readerGroup = request.getRgName();
+        long requestId = request.getRequestId();
+        UUID readerGroupId = request.getReaderGroupId();
+        ImmutableSet<String> streamsToBeUnsubscribed = request.getRemoveStreams();
+        final RGOperationContext context = streamMetadataStore.createRGContext(scope, readerGroup);
+
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getReaderGroupId(scope, readerGroup, context, executor)
+                .thenCompose(id -> {
+                if (!id.equals(readerGroupId)) {
+                        log.warn("Skipping processing of Reader Group update request {} as UUIDs did not match.", requestId);
+                        return CompletableFuture.completedFuture(null);
+                }
+                return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
+                       .thenCompose(rgConfigRecord -> {
+                       if (rgConfigRecord.getObject().isUpdating() &&
+                          (!ReaderGroupConfig.StreamDataRetention.values()[rgConfigRecord.getObject().getRetentionTypeOrdinal()]
+                            .equals(ReaderGroupConfig.StreamDataRetention.NONE))) {
+                            // update Stream metadata tables, only if RG is a Subscriber
+                            Iterator<String> streamIter = rgConfigRecord.getObject().getStartingStreamCuts().keySet().iterator();
+                            String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
+                            Iterator<String> removeStreamsIter = streamsToBeUnsubscribed.stream().iterator();
+                            return Futures.loop(() -> streamIter.hasNext(), () -> {
+                                          Stream stream = Stream.of(streamIter.next());
+                                          return streamMetadataStore.addSubscriber(stream.getScope(),
+                                                       stream.getStreamName(), scopedRGName, rgConfigRecord.getObject().getGeneration(),
+                                                  null, executor);
+                                      }, executor)
+                                    .thenCompose(v -> Futures.loop(() -> removeStreamsIter.hasNext(), () -> {
+                                        Stream stream = Stream.of(removeStreamsIter.next());
+                                        return streamMetadataStore.deleteSubscriber(stream.getScope(),
+                                                stream.getStreamName(), scopedRGName, null, executor);
```

Comment:
shouldnt generation be included in deleteSubscriber too? we dont want to be deleting an unintended subscriber just because this event got replayed.

Reply:
The UUID check in deleteRGEvent should take care of that, but there is no harm in doing the check again in deleteSubscriber, so will add it here.

Reply:
fixed